File: java/day25_part1_2022.java

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) throws IOException {
        long sum = 0;
        Scanner scanner = new Scanner(new File("input.txt"));
        while (scanner.hasNextLine()) {
            sum += fromSnafu(scanner.nextLine());
        }
        System.out.println(toSnafu(sum));
        scanner.close();
    }

    private static long fromSnafu(String s) {
        long n = 0;
        for (int i = 0; i < s.length(); i++) {
            n *= 5;
            switch (s.charAt(i)) {
                case '=':
                    n -= 2;
                    break;
                case '-':
                    n--;
                    break;
                default:
                    n += s.charAt(i) - '0';
            }
        }
        return n;
    }

    private static String toSnafu(long n) {
        if (n == 0) return "0";
        StringBuilder sb = new StringBuilder();
        while (n > 0) {
            switch ((int)(n % 5)) {
                case 3:
                    n += 5;
                    sb.append('=');
                    break;
                case 4:
                    n += 5;
                    sb.append('-');
                    break;
                default:
                    sb.append((char)('0' + n % 5));
            }
            n /= 5;
        }
        return sb.reverse().toString();
    }
}

File: csharp/day25_part1_2022.cs

using System;
using System.IO;
using System.Text;

class Program
{
    static void Main()
    {
        var lines = File.ReadAllLines("input.txt");
        long sum = 0;
        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0) continue;
            sum += FromSnafu(trimmed);
        }
        Console.WriteLine(ToSnafu(sum));
    }

    static long FromSnafu(string s)
    {
        long n = 0;
        foreach (char c in s)
        {
            n *= 5;
            if (c == '=') n -= 2;
            else if (c == '-') n--;
            else n += c - '0';
        }
        return n;
    }

    static string ToSnafu(long n)
    {
        if (n == 0) return "0";
        var sb = new StringBuilder();
        while (n > 0)
        {
            int r = (int)(n % 5);
            if (r == 3)
            {
                n += 5;
                sb.Append('=');
            }
            else if (r == 4)
            {
                n += 5;
                sb.Append('-');
            }
            else
            {
                sb.Append((char)('0' + r));
            }
            n /= 5;
        }
        var res = new char[sb.Length];
        for (int i = 0; i < sb.Length; i++) res[i] = sb[sb.Length - 1 - i];
        return new string(res);
    }
}
